File: src/main/java/com/github/cao/awa/kalmia/message/unknown/UnknownMessage.java
package com.github.cao.awa.kalmia.message.unknown;

import com.github.cao.awa.apricot.io.bytes.reader.BytesReader;
import com.github.cao.awa.apricot.util.digger.MessageDigger;
import com.github.cao.awa.kalmia.annotations.auto.network.unsolve.AutoData;
import com.github.cao.awa.kalmia.env.KalmiaEnv;
import com.github.cao.awa.kalmia.mathematic.Mathematics;
import com.github.cao.awa.kalmia.message.Message;
import com.github.cao.awa.kalmia.message.digest.DigestData;

public class UnknownMessage extends Message {
    @AutoData
    private byte[] msgBytes;
    @AutoData
    private DigestData digest;

    public UnknownMessage() {

    }

    public UnknownMessage(byte[] msgBytes) {
        this.msgBytes = msgBytes;
        this.digest = new DigestData(MessageDigger.Sha3.SHA_512,
                                     Mathematics.toBytes(MessageDigger.digest(msgBytes,
                                                                              MessageDigger.Sha3.SHA_512
                                                         ),
                                                         16
                                     )
        );
    }

    @Override
    public DigestData digest() {
        return this.digest;
    }

    public byte[] details() {
        return this.msgBytes;
    }

    @Override
    public long getSender() {
        return - 1;
    }

    public static UnknownMessage create(BytesReader reader) {
        try {
            return KalmiaEnv.BYTES_SERIALIZE_FRAMEWORK.create(new UnknownMessage(),
                                                              reader
            );
        } catch (Exception e) {
            e.printStackTrace();
            throw new RuntimeException(e);
        }
//        if (reader.read() == 0) {
//            String msg;
//
//            long sender = SkippedBase256.readLong(reader);
//
//            int msgLength = Base256.tagFromBuf(reader.read(2));
//            msg = new String(reader.read(msgLength),
//                             StandardCharsets.UTF_8
//            );
//
//            DigestData digestData = DigestData.create(reader);
//
//            return new PlainsMessage(msg,
//                                     sender,
//                                     digestData
//            );
//        } else {
//            return null;
//        }
    }
}

File: src/main/java/com/github/cao/awa/kalmia/message/unknown/UnknownMessage.kt
package com.github.cao.awa.kalmia.message.unknown

import com.github.cao.awa.apricot.annotations.auto.Auto
import com.github.cao.awa.apricot.identifier.BytesRandomIdentifier
import com.github.cao.awa.apricot.io.bytes.reader.BytesReader
import com.github.cao.awa.apricot.util.digger.MessageDigger
import com.github.cao.awa.kalmia.annotations.auto.network.unsolve.AutoData
import com.github.cao.awa.kalmia.identity.LongAndExtraIdentity
import com.github.cao.awa.kalmia.mathematic.Mathematics
import com.github.cao.awa.kalmia.message.Message
import com.github.cao.awa.kalmia.message.digest.DigestData
import com.github.cao.awa.kalmia.message.display.ClientMessageContent
import com.github.cao.awa.viburnum.util.bytes.BytesUtil

class UnknownMessage : Message {
    companion object {
        private val HEADER = byteArrayOf(0)

        @JvmStatic
        fun create(reader: BytesReader): UnknownMessage? {
            return if (reader.read().toInt() == 0) {
                val identity = LongAndExtraIdentity.read(reader)
                val data = reader.all()
                UnknownMessage(
                    identity,
                    data
                )
            } else {
                null
            }
        }
    }

    @AutoData
    private var msgBytes: ByteArray

    @AutoData
    private var digestData: DigestData

    @Auto
    constructor() {
        this.msgBytes = BytesUtil.EMPTY
        this.digestData = DigestData()
    }

    constructor(msgBytes: ByteArray) {
        this.msgBytes = msgBytes
        this.digestData = DigestData.digest(
            MessageDigger.Sha3.SHA_512,
            msgBytes
        )
    }

    constructor(identity: LongAndExtraIdentity, msgBytes: ByteArray) : super(identity) {
        this.msgBytes = msgBytes
        this.digestData = DigestData.digest(
            MessageDigger.Sha3.SHA_512,
            msgBytes
        )
    }

    override fun digest(): DigestData {
        return this.digestData
    }

    fun details(): ByteArray {
        return this.msgBytes
    }

    override fun sender(): LongAndExtraIdentity {
        return LongAndExtraIdentity.create(-1, BytesRandomIdentifier.create(16))
    }

    override fun header(): ByteArray {
        return HEADER
    }

    override fun display(): ClientMessageContent {
        return ClientMessageContent(
            sender(),
            "PlainsMessage{sender=${sender()}, msg=${Mathematics.radix(details(), 36)}, digest=${digest().value36()}}",
            """
                This message unable to display, it digest is:
                ${digest().value36()}
                sender of this message is:
                ${sender()}
                it has theres data: 
                ${Mathematics.radix(details(), 36)}
                """
        )
    }

    override fun toBytes(): ByteArray {
        return BytesUtil.concat(
            header(),
            identity().toBytes(),
            details()
        )
    }
}